File: tools/generate-config.php
<?php
// phpcs:ignoreFile

if(php_sapi_name() !== 'cli') {
    echo 'This script can only be run from the command line.';
    exit(1);
}

if($argc < 2) {
    echo "Usage: php generate-config.php h|c|conf\n";
    exit(1);
}

for($i = 1; $i < $argc; ++$i) {
    if(!in_array($argv[1], ['h', 'c', 'conf', 'server'])) {
        echo "Wrong file type. Accepted files: h|c|conf|server\n";
        exit(1);
    }
}

// List of configurations.
$map = [
    'Database',
    ['db_path', 'file:warudo.db'],
    ['log_level', 0, ['no_log', 'info', 'error', 'debug'], 'Log level can be one of the following [0, 1, 2, 3]: no_log, info, error, debug'],
    'Net',
    ['access_origin', null],
    ['listen_backlog', 1024],
    ['max_columns', 64],
    ['net_buffer_size', 1],
    ['net_headers_buffer_size', 4096],
    ['net_input_buffer_size', 1],
    ['socket_port', 6251],
    ['timing', 1],
];

// Configuration file.
$confs = [];

// H header file.
$defines = [];
$names = [];
$statuses = [];
$structs = [];

// C file.
$init_configs = [];
$free_configs = [];
$db_loads = [];
$env_loads = [];

// Server file.
$logs = [];

$struct_count = count(array_filter($map, 'is_array'));
$statuses[] = "// Configuration statuses.";
$structs[] = "char config_status[{$struct_count}];";
$structs[] = '';
$init_configs[] = '// Initialize to defaults.';
$init_configs[] = "memset(config->config_status, WRD_DEFAULT_CONFIG, {$struct_count});\n";

// Loop through each line
foreach($map as $value) {
    $comment = null;

    if(is_string($value)) {
        $comment = $value;
    }

    // A section
    if(is_string($value)) {
        $comment = "// {$value}";
        // Configuration file.
        $confs[] = $comment;

        // H file.
        $defines[] = "\n" . $comment;
        $structs[] = $comment;

        // C file.
        $init_configs[] = $comment;
        $free_configs[] = $comment;
        $env_loads[] = $comment;

        continue;
    }

    if(!is_array($value) || !count($value)) {
        continue;
    }

    if(count($value) === 4) {
        $confs[] = '// ' . $value[3];
    }

    $entry_name = $value[0];
    $entry_value = $value[1];
    $enum_values = $value[2] ?? [];
    $is_integer = is_int($entry_value);
    $define_name = strtoupper($entry_name);
    $define = 'WRD_DEFAULT_' . $define_name;
    $type = $is_integer ? 'int ' : 'char *';
    $env_function = $is_integer ? 'LOAD_ENV_CONFIG_INT' : 'LOAD_ENV_CONFIG_STRING';
    $env_cast = $is_integer ? '(int*)' : '';

    // Configuration file.
    $conf_entry_value = value_to_conf($entry_value);
    $confs[] = "{$entry_name}: {$conf_entry_value},";

    // An enum
    if(count($enum_values)) {
        $type = 'wrd_' . $entry_name;
        $defines[] = 'typedef enum {';
        $enums = [];

        foreach($enum_values as $enum_value) {
            $enums[] = 'WRD_' . strtoupper($entry_name) . '_' . strtoupper( $enum_value );
        }

        $defines[] = join(",\n", add_indentation($enums));
        $defines[] = "} {$type};\n";
        $type .= ' ';
    }

    $c_entry_value = value_to_c($entry_value);

    // H file.
    $defines[] = "#define {$define} " . $c_entry_value;
    $structs[] = "{$type}{$entry_name};";
    $names[] = "WRD_{$define_name},";

    // C file.
    $init_configs[] = "config->{$entry_name} = {$define};";
    $config_name = explode('_', $entry_name);
    $config_name = array_map('ucfirst', $config_name);
    $config_name = join(' ', $config_name);

    if($is_integer) {
        $db_loads[] = "LOAD_DB_CONFIG_INT(WRD_{$define_name}, {$entry_name})";
        $logs[] = "wrd_log_info(config, u8\"{$config_name}: %d [%c]\\n\", config->{$entry_name}, wrd_get_config_status(config, WRD_{$define_name}));";
    } else {
        $db_loads[] = "LOAD_DB_CONFIG_STR(WRD_{$define_name}, {$entry_name})";
        $logs[] = "wrd_log_info(config, u8\"{$config_name}: %s [%c]\\n\", config->{$entry_name}, wrd_get_config_status(config, WRD_{$define_name}));";
    }

    if(!$is_integer) {
        // $free_configs[] = "if(config->{$entry_name}) {\n        free(config->{$entry_name});\n        config->{$entry_name} = NULL;\n    }\n";
        $free_configs[] = "FREE_CONFIG_STRING(WRD_{$define_name}, $entry_name)\n";
    }

    $env_loads[] = "{$env_function}(WRD_{$define_name}, {$entry_name})";
}

$defines = array_merge(
    [
        '// Configuration names.',
        'typedef enum {'
    ],
    add_indentation($names),
    ['} wrd_config_name;'],
    $defines
);

$files = [
    'h' => [
        'file' => 'src/warudo.h',
        'start' => '// Configurations.',
        'end' => "\n#include <",
        'text' => join("\n", $defines) . "\n",
        'additionals' => [
            [
                'end' => "\n    // Network",
                'text' => "\n" . join("\n", add_indentation($structs)) . "\n",
            ],
        ]
    ],
    'c' => [
        'file' => 'src/conf.c',
        'start' => '// Configurations.',
        'end' => "\n    return WRD_OK;",
        'text' => "\n" . join("\n", add_indentation($init_configs)) . "\n",
        'additionals' => [
            [
                'end' => "\n    return WRD_OK;",
                'text' => "\n" . join("\n", add_indentation($env_loads)) . "\n",
            ], [
                'end' => "\n    return WRD_OK;",
                'text' => "\n" . join("\n", add_indentation($free_configs)),
            ], [
                'end' => "\n    wrd_load_config_env",
                'text' => "\n" . join("\n", add_indentation($db_loads)) . "\n",
            ]
        ]
    ],
    'conf' => [
        'file' => 'warudo.conf.default',
        'start' => '{',
        'end' => '}',
        'text' => join("\n", add_indentation($confs)) . "\n",
        'additionals' => [],
    ],
    'server' => [
        'file' => 'src/server.c',
        'start' => '// Configurations.',
        'end' => "res = wrd_db_init(config->db_path, config);",
        'text' => join("\n", add_indentation($logs)) . "\n\n    ",
        'additionals' => [],
    ],
];

for($i = 1; $i < $argc; ++$i) {
    $type = $argv[$i];
    $file = $files[$type];
    $content = file_get_contents($file['file']);
    $injection = inject_text($content, $file['start'], $file['end'], $file['text']);

    foreach($file['additionals'] as $additional) {
        $injection = inject_text($injection['text'], $file['start'], $additional['end'], $additional['text'], $injection['start_pos']);
    }

    if(file_put_contents($file['file'], $injection['text']) === false) {
        echo "Could not write to {$file['file']}.\n";
        exit(1);
    }

    echo "Generated {$file['file']}.\n";
}

// Transform a value to a configuration file value.
function value_to_conf($value) {
    if(is_null($value)) {
        return 'null';
    }

    if(is_int($value)) {
        return $value;
    }

    return '\'' . str_replace('\'', '\\\'', $value) . '\'';
}

// Transform a value to a C file value.
function value_to_c($value) {
    if(is_null($value)) {
        return 'NULL';
    }

    if(is_int($value)) {
        return $value;
    }

    return '"' . str_replace('"', '\\"', $value) . '"';
}

// Add indentation to an array of values.
function add_indentation($values, $indentation = 1) {
    $indentation = str_repeat(' ', $indentation * 4);

    return array_map(
        function($value) use ($indentation) {
            return $indentation . $value;
        },
        $values
    );
}

function inject_text($file, $start_text, $end_text, $text, $start_pos = 0) {
    $start_pos = strpos($file, $start_text, $start_pos);

    if($start_pos === false) {
        echo "Could not find the start of injection point ('{$start_text}').\n";
        exit(1);
    }

    $start_pos += strlen($start_text) + 1;
    $end_pos = strpos($file, $end_text, $start_pos);

    if($end_pos === false) {
        echo "Could not find the end of the injection point ('{$end_text}').\n";
        exit(1);
    }

    return [
        'start_pos' => $start_pos,
        'text' => substr_replace($file, $text, $start_pos, $end_pos - $start_pos),
    ];
}

File: tools/generate-config.py
from sys import argv
import utils as ut

if len(argv) < 2:
    print("Usage: python3 generate-config.py h|c|conf")
    exit(1)

type Configuration = list[str | map]

cfg_map: Configuration = ut.get_data_from_json_file("tools/configs.json")

# Configuration file.
confs: list[str] = []

# H header file.
defines: list[str] = []
names: list[str] = []
statuses: list[str] = []
structs: list[str] = []

# C file.
init_configs: list[str] = []
free_configs: list[str] = []
db_loads: list[str] = []
env_loads: list[str] = []

# Server file.
logs: list[str] = []

struct_count = sum(isinstance(item, dict) for item in cfg_map)
statuses.append("// Configuration statuses.")
structs.append(f"char config_status[{struct_count}];\n")
init_configs.append("// Initialize to defaults.")
init_configs.append(
    f"memset(config->config_status, WRD_DEFAULT_CONFIG, {struct_count});\n"
)

# Loop configs.json.
for value in cfg_map:
    comment = None

    if isinstance(value, str):
        comment = f"// {value}"

        # Configuration file.
        confs.append(comment)

        # h file.
        defines.append(f"\n{comment}")
        structs.append(comment)

        # c file.
        init_configs.append(comment)
        free_configs.append(comment)
        env_loads.append(comment)

        continue

    if not isinstance(value, dict):
        continue

    if "comment" in value:
        confs.append(f"// {value["comment"]}")

    entry_name = value["name"]
    entry_value = value["default"]
    enum_values = value.get("enum", [])
    is_int = isinstance(entry_value, int) and not value.get("type") == "mixed"
    define_name = entry_name.upper()
    define = f"WRD_DEFAULT_{define_name}"
    has_enum = len(enum_values)

    if is_int:
        entry_type = "int "
        env_function = "LOAD_ENV_CONFIG_INT"
        env_cast = "(int*)"
    else:
        entry_type = "char *"
        env_function = "LOAD_ENV_CONFIG_STRING"
        env_cast = ""

    # Configuration file.
    conf_entry_value = ut.value_to_conf(entry_value)
    confs.append(f"{entry_name}: {conf_entry_value},")

    # An enum.
    if has_enum:
        entry_type = f"wrd_{entry_name}"
        defines.append("typedef enum {")
        enums = []

        for enum_value in enum_values:
            enums.append(f"WRD_{entry_name.upper()}_{enum_value.upper()}")

        defines.append(",\n".join(ut.add_indent(enums)))
        defines.append(f"}} {entry_type};\n")
        entry_type += " "

    c_entry_value = ut.value_to_c(entry_value, is_int)
    define_name = f"WRD_{define_name}"

    # h file.
    if has_enum:
        defines.append(f"#define {define} {enums[c_entry_value]}")
    else:
        defines.append(f"#define {define} {c_entry_value}")

    structs.append(f"{entry_type}{entry_name};")
    names.append(f"{define_name},")

    # c file.
    init_configs.append(f"config->{entry_name} = {define};")
    config_name = entry_name.split("_")
    config_name = [s.capitalize() for s in config_name]
    config_name = " ".join(config_name)

    if is_int:
        db_loads.append(f"LOAD_DB_CONFIG_INT({define_name}, {entry_name})")
        logs.append(
            f'wrd_log_info(config, u8"{config_name}: %d [%c]\\n", '
            f"config->{entry_name}, wrd_get_config_status(config, "
            f"{define_name}));"
        )
    else:
        db_loads.append(f"LOAD_DB_CONFIG_STR({define_name}, {entry_name})")
        logs.append(
            f'wrd_log_info(config, u8"{config_name}: %s [%c]\\n", '
            f"config->{entry_name}, wrd_get_config_status(config, "
            f"{define_name}));"
        )

    if not is_int:
        free_statement = f"FREE_CONFIG_STRING({define_name}, {entry_name})"
        free_configs.append(free_statement)

    env_loads.append(f"{env_function}({define_name}, {entry_name})")

defines: list[str] = (
    ["// Configuration names.", "typedef enum {"]
    + ut.add_indent(names)
    + ["} wrd_config_name;"]
    + defines
)

type File = dict[str, str | list[dict[str, str]]]

files: File = {
    "h": {
        "file": "src/warudo.h",
        "start": "// Configurations.",
        "end": "\n#include <",
        "text": "\n".join(defines) + "\n",
        "additionals": [
            {
                "end": "\n    // Network",
                "text": "\n" + "\n".join(ut.add_indent(structs)) + "\n",
            },
        ],
    },
    "c": {
        "file": "src/conf.c",
        "start": "// Configurations.",
        "end": "\n    return WRD_OK;",
        "text": "\n" + "\n".join(ut.add_indent(init_configs)) + "\n",
        "additionals": [
            {
                "end": "\n    return WRD_OK;",
                "text": "\n" + "\n".join(ut.add_indent(env_loads)) + "\n",
            },
            {
                "end": "\n    return WRD_OK;",
                "text": "\n" + "\n".join(ut.add_indent(free_configs)) + "\n",
            },
            {
                "end": "\n    wrd_load_config_env",
                "text": "\n" + "\n".join(ut.add_indent(db_loads)) + "\n",
            },
        ],
    },
    "conf": {
        "file": "warudo.conf.default",
        "start": "{",
        "end": "}",
        "text": "\n".join(ut.add_indent(confs)) + "\n",
        "additionals": [],
    },
    "server": {
        "file": "src/server.c",
        "start": "// Configurations.",
        "end": "ret = wrd_db_init(config->db_path, config);",
        "text": "\n".join(ut.add_indent(logs)) + "\n\n    ",
        "additionals": [],
    },
}

for i, arg in enumerate(argv[1:]):
    f = files[arg]

    with open(f["file"], "r") as file:
        content = file.read()

    file.close()
    res = ut.generate_content(content, f["start"], f["end"], f["text"])

    for additional in f["additionals"]:
        res = ut.generate_content(
            res["text"],
            f["start"],
            additional["end"],
            additional["text"],
            res["start_pos"],
        )

    with open(f["file"], "w") as file:
        file.write(res["text"])

    print(f"Generated {f['file']}")
